lib/pure/random.nim

```diff
@@ -277,9 +278,32 @@ proc rand*[T: Ordinal or SomeFloat](max: T): T {.benign.} =
     doAssert rand(100) == 96
     doAssert rand(100) == 66
 
+  result = rand(state, int(max))
+
+proc rand*[T: SomeFloat](max: T): float {.benign.} =
```

Comment:
I don't have a good answer to how to fix this, but:
* what happens on a 32bit machine when you call `rand(maximumPositiveValue(float64))` or similar?
* what happens on a 32bit machine when you call `rand(int64(int32.high) + 1)` or similar?
* also on a 64bit machine, things like `echo rand(uint64.high-1)` fail with `RangeDefect`

the problem is that returning `int` or `float` seems flawed; I don't know if it's fixable or not, and whether `rand(max:T)` should be deprecated.

Note that `rand(HSlice)` and `rand(typedesc[T])` are much more sound and dont' have those issues and can always be used as workaround

Reply:
=> https://github.com/nim-lang/Nim/pull/16291#issuecomment-740913908